File: python/src/ev3/ut6_pooavanzadaestructuras/u01herenciapolimorfismo/practicas/madridcesjuanpablosegundotrabajadores/trabajadores_resuelto.py
"""Trabajadores CES — espejo Kotlin."""
from abc import ABC, abstractmethod


class Trabajador(ABC):
    def __init__(self, nombre: str, apellido: str, dni: str, sueldo: float):
        self.nombre = nombre
        self.apellido = apellido
        self.dni = dni
        self.sueldo = sueldo

    @abstractmethod
    def calcular_salario_mensual(self) -> float:
        ...

    def mostrar_datos(self) -> None:
        print(f"nombre = {self.nombre}")
        print(f"apellido = {self.apellido}")
        print(f"dni = {self.dni}")
        print(f"sueldo = {self.sueldo}")
        print(f"salario mensual = {self.calcular_salario_mensual()}")


class Asalariado(Trabajador):
    def __init__(self, nombre, apellido, dni, sueldo, numero_pagas, contratado, cuota_ss):
        super().__init__(nombre, apellido, dni, sueldo)
        self.numero_pagas = numero_pagas
        self.contratado = contratado
        self.cuota_ss = cuota_ss

    def calcular_salario_mensual(self) -> float:
        return (self.sueldo * 0.79) / self.numero_pagas

    def mostrar_datos(self) -> None:
        print("Tipo = Asalariado")
        super().mostrar_datos()


class Autonomo(Trabajador):
    def __init__(self, nombre, apellido, dni, sueldo, contratado, cuota_ss):
        super().__init__(nombre, apellido, dni, sueldo)
        self.contratado = contratado
        self.cuota_ss = cuota_ss

    def calcular_salario_mensual(self) -> float:
        return (self.sueldo - 12 * self.cuota_ss) / 12

    def mostrar_datos(self) -> None:
        print("Tipo = Autonomo")
        super().mostrar_datos()


class Jefe(Trabajador):
    def __init__(self, nombre, apellido, dni, sueldo, acciones, beneficio):
        super().__init__(nombre, apellido, dni, sueldo)
        self.acciones = acciones
        self.beneficio = beneficio

    def calcular_salario_mensual(self) -> float:
        return self.sueldo / 12

    def mostrar_datos(self) -> None:
        print("Tipo = Jefe")
        super().mostrar_datos()


def main() -> None:
    lista: list[Trabajador] = []
    opcion = 0
    while opcion != 7:
        print("1 Asalariado 2 Autonomo 3 Jefe 4 Listar 5 PorDNI 6 Despedir 7 Salir")
        opcion = int(input("Opcion: "))
        if opcion == 1:
            lista.append(Asalariado(
                input("nombre: "), input("apellido: "), input("dni: "),
                float(input("sueldo: ")), int(input("pagas: ")), True, float(input("cuota: "))
            ))
        elif opcion == 2:
            lista.append(Autonomo(
                input("nombre: "), input("apellido: "), input("dni: "),
                float(input("sueldo: ")), True, float(input("cuota: "))
            ))
        elif opcion == 3:
            lista.append(Jefe(
                input("nombre: "), input("apellido: "), input("dni: "),
                float(input("sueldo: ")), int(input("acciones: ")), float(input("beneficio: "))
            ))
        elif opcion == 4:
            for t in lista:
                t.mostrar_datos()
        elif opcion == 5:
            dni = input("DNI: ")
            for t in lista:
                if t.dni.lower() == dni.lower():
                    t.mostrar_datos()
        elif opcion == 6:
            dj = input("DNI jefe: ")
            dt = input("DNI trabajador: ")
            if not any(isinstance(t, Jefe) and t.dni.lower() == dj.lower() for t in lista):
                print("El DNI del jefe no es correcto")
            else:
                lista[:] = [t for t in lista if not (t.dni.lower() == dt.lower() and not isinstance(t, Jefe))]
                print("Trabajador despedido")
        elif opcion == 7:
            print("Saliendo")


if __name__ == "__main__":
    main()

File: csharp/src/ev3/ut6_pooavanzadaestructuras/u01herenciapolimorfismo/practicas/madridcesjuanpablosegundotrabajadores/TrabajadoresResuelto.cs
using System; using System.Collections.Generic; using System.Linq;
abstract class Trabajador{ public string Nombre,Apellido,Dni; public double Sueldo;
  public Trabajador(string n,string a,string d,double s){Nombre=n;Apellido=a;Dni=d;Sueldo=s;}
  public abstract double CalcularSalarioMensual();
  public virtual void MostrarDatos(){ Console.WriteLine($"nombre = {Nombre}"); Console.WriteLine($"dni = {Dni}"); Console.WriteLine($"salario mensual = {CalcularSalarioMensual()}"); } }
class Asalariado:Trabajador{ public int Pagas; public double Cuota; public Asalariado(string n,string a,string d,double s,int p,double c):base(n,a,d,s){Pagas=p;Cuota=c;}
  public override double CalcularSalarioMensual()=>(Sueldo*0.79)/Pagas; public override void MostrarDatos(){Console.WriteLine("Tipo = Asalariado"); base.MostrarDatos();} }
class Autonomo:Trabajador{ public double Cuota; public Autonomo(string n,string a,string d,double s,double c):base(n,a,d,s){Cuota=c;}
  public override double CalcularSalarioMensual()=>(Sueldo-12*Cuota)/12; public override void MostrarDatos(){Console.WriteLine("Tipo = Autonomo"); base.MostrarDatos();} }
class Jefe:Trabajador{ public int Acciones; public double Beneficio; public Jefe(string n,string a,string d,double s,int ac,double b):base(n,a,d,s){Acciones=ac;Beneficio=b;}
  public override double CalcularSalarioMensual()=>Sueldo/12; public override void MostrarDatos(){Console.WriteLine("Tipo = Jefe"); base.MostrarDatos();} }
class TrabajadoresResuelto{ static void Main(){ var lista=new List<Trabajador>(); int op; do{
  Console.WriteLine("1 Asalariado 2 Autonomo 3 Jefe 4 Listar 5 DNI 6 Despedir 7 Salir"); op=int.Parse(Console.ReadLine()!);
  if(op==1) lista.Add(new Asalariado(Console.ReadLine()!,Console.ReadLine()!,Console.ReadLine()!,double.Parse(Console.ReadLine()!),int.Parse(Console.ReadLine()!),double.Parse(Console.ReadLine()!)));
  else if(op==2) lista.Add(new Autonomo(Console.ReadLine()!,Console.ReadLine()!,Console.ReadLine()!,double.Parse(Console.ReadLine()!),double.Parse(Console.ReadLine()!)));
  else if(op==3) lista.Add(new Jefe(Console.ReadLine()!,Console.ReadLine()!,Console.ReadLine()!,double.Parse(Console.ReadLine()!),int.Parse(Console.ReadLine()!),double.Parse(Console.ReadLine()!)));
  else if(op==4) foreach(var t in lista) t.MostrarDatos();
  else if(op==5){ var dni=Console.ReadLine()!; foreach(var t in lista) if(t.Dni.Equals(dni,StringComparison.OrdinalIgnoreCase)) t.MostrarDatos(); }
  else if(op==6){ var dj=Console.ReadLine()!; var dt=Console.ReadLine()!;
    if(!lista.Any(t=>t is Jefe && t.Dni.Equals(dj,StringComparison.OrdinalIgnoreCase))) Console.WriteLine("El DNI del jefe no es correcto");
    else { lista.RemoveAll(t=>t.Dni.Equals(dt,StringComparison.OrdinalIgnoreCase) && t is not Jefe); Console.WriteLine("Trabajador despedido"); } }
} while(op!=7);} }
